### PR TITLE
(fix) Week numbers may break at end/begin of the years

### PR DESCRIPTION
**Pull request by tjfred35**
Fixes some weird edge cases where weeks at the beginning or end of the year could render in the wrong order, or share the same table row.

This fixes the following issues:
https://github.com/gpbl/react-day-picker/issues/1207
https://github.com/gpbl/react-day-picker/issues/1222
https://github.com/gpbl/react-day-picker/issues/1250

The problem came from using week number in a dictionary, because at the beginning and end of years, the natural order of a date-fns week number is not the order that we want to render the weeks in.

So instead of returning an object, we return an array of a new MonthWeek object from getWeeks function, to maintain order. Furthermore, we remove the previous attempt to fix this ordering problem that was the source of one of the above issues where two weeks would render into one row. By removing this week transformation code, we can reduce the overall complexity of what a week number is.

### CODE REVIEW

File: packages/react-day-picker/src/components/Table/Table.test.tsx [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+const po = new PageObjects(new Date());
+
+describe('It should render weeks at the boundaries of years correctly', () => {
+  test('If the first days of the month start at week 52 from the previous year, week 52 should be rendered on the 1st row', () => {
+    customRender(<Table displayMonth={new Date(2022, 0)} />, { locale: enGB });
+    const week1 = po.weeks[0].querySelectorAll('td');
+    const day6 = week1[5];
+    const day7 = week1[6];
+    expect(day6).toHaveTextContent('11st January (Saturday)');
+    expect(day7).toHaveTextContent('22nd January (Sunday)');
```
Comment (tjfred35) [inline packages/react-day-picker/src/components/Table/Table.test.tsx:17]: I would love suggestions on how to structure these tests better -- right now the text content is from an aria-hidden element (`"2"`) and hidden element that is presumably read by screen readers (`"2nd January (Sunday)"`) -- I don't know to what extend that construct in this v8 beta is accessible.

File: packages/react-day-picker/src/components/Table/Table.test.tsx [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+  test('If the last week of the year is 1, it should be rendered on the 5th row', () => {
+    customRender(<Table displayMonth={new Date(2021, 11)} />, {
+      locale: enUS
+    });
+    const week5 = po.weeks[4].querySelectorAll('td');
+    const day29 = week5[0];
+    const day30 = week5[1];
+    expect(day29).toHaveTextContent('2626th December (Sunday)');
+    expect(day30).toHaveTextContent('2727th December (Monday)');
```
Comment (tjfred35) [inline packages/react-day-picker/src/components/Table/Table.test.tsx:41]: This test passes with the previous code because of this code:
```
if (week === 1 && getMonth(date) === 11) {
  week = 53;
}
```
With the proposed code changes, the above code has been removed, but the test still passes, and the other two edge case tests do too.

However, removing that code above now renders the last month of the year with a week number of one (see screenshot):
Screen Shot 2021-11-02 at 3 09 22 PM

I am making the assumption that the reason for the code above was to fix the original ordering bug, not to apply any desired transformations to the week numbers.

The week number, 1, without the transformation to map it to 53, is now consistent between the December 2021 and the January 2022 calendars, so I think it's a reasonable change of behavior to the rendering of the week numbers that also fixes these other edge cases. Additionally, making week number transformations significantly increases the complexity of the definition of a week number, so removing the transformation is an improvement as far as reducing complexity is concerned.

File: packages/react-day-picker/src/components/Table/utils/getWeeks.ts [hunk 3]
```diff
@@ -14,51 +13,44 @@ import { getOutsideEndDays } from './getOutsideEndDays';
 import { getOutsideStartDays } from './getOutsideStartDays';
 
 /**
- * The weeks belonging to a month. Each key of the returned object is the
- * week number of the year.
+ * The date-fns week and the dates for that week
  */
-export type MonthWeeks = { [weeknumber: string]: Date[] };
-
+export type MonthWeek = { weekNumber: number; dates: Date[] };
```
Comment (tjfred35) [inline packages/react-day-picker/src/components/Table/utils/getWeeks.ts:18]: Since this type and the `getWeeks` function were previously publicly exported, this change is a breaking change, but I assumed that is ok because we are still in beta.

File: packages/react-day-picker/src/components/Table/Table.test.tsx (hunk 1 above)
Comment (UlyssesInvictus) [reply packages/react-day-picker/src/components/Table/Table.test.tsx:17]: Is there a reason the text of the actual rendered dates (the non-hidden elements) isn't preferred here? Are they not actually visible?

Also, is this react testing library?

File: packages/react-day-picker/src/components/Table/Table.test.tsx (hunk 2 above)
Comment (UlyssesInvictus) [reply packages/react-day-picker/src/components/Table/Table.test.tsx:41]: As long as it's documented in the (eventual) library docs, I don't see that as a breaking change. It's a lot easier to understand "week # = latest week # within this 7 day period" than the edge case of the end of the year.

The only thing that gives me pause is that a week 53 is extremely rare but technically possible. However, the number 53 isn't special in itself. Are there any known use cases where someone has to detect whether something is the last week, and is currently reliant on looking for a hardcoded value of 52/53 ?

File: packages/react-day-picker/src/components/Table/Table.test.tsx (hunk 1 above)
Comment (melinayoung-toast) [reply packages/react-day-picker/src/components/Table/Table.test.tsx:17]: In addition to these tests, have you thought about including snapshot tests? They can be used to verify that the behavior doesn't change in the future https://jestjs.io/docs/snapshot-testing

Comment (jericopingul-toast) [reply packages/react-day-picker/src/components/Table/Table.test.tsx:17]: One way to select the cell in a particular role is something like
```js
const table = screen.getByRole('table')
const firstBodyRow = within(table).getAllByRole('row')[2]
within(firstBodyRow).getByRole('cell', {
  name: /3rd january \(monday\)/i
})
```
Since there are a lot of nesting `within` can be useful:
https://testing-library.com/docs/dom-testing-library/api-within/

Comment (jericopingul-toast) [reply packages/react-day-picker/src/components/Table/Table.test.tsx:17]: This might also work if you want to select a row with a matching date
```js
const table = screen.getByRole('table')
within(table).getByRole('row', {
  name: /3rd january \(monday\)/i
})
```
The testing playground is a pretty nice tool to play around with selecting elements and gives recommendations
https://testing-library.com/docs/queries/about/#screenlogtestingplaygroundurl

Comment (tjfred35) [reply packages/react-day-picker/src/components/Table/Table.test.tsx:17]: Yes -- but then how to assert that particular element is in the right position in the grid?

Comment (jericopingul-toast) [reply packages/react-day-picker/src/components/Table/Table.test.tsx:17]: Good point, can we use something like the the row/col indexes?
```js
const table = screen.getByRole('table')
const firstBodyRow = within(table).getAllByRole('row')[2]
const targetDate = within(firstBodyRow).getAllByRole('cell')[0]
```
There might be a better way than this...

File: packages/react-day-picker/.eslintrc [hunk 4]
```diff
@@ -4,6 +4,9 @@
     "ecmaVersion": 2018,
     "sourceType": "module"
   },
+  "env": {
+    "jest": true
+  },
```
Comment (gpbl) [inline packages/react-day-picker/.eslintrc:9]: There's already an override for the test files. This should not be needed.

```suggestion
```

File: packages/react-day-picker/src/components/Table/utils/getWeeks.ts [hunk 5]
```diff
@@ -14,51 +13,44 @@ import { getOutsideEndDays } from './getOutsideEndDays';
...
+  const weeks: MonthWeek[] = [];
+  let lastWeek: MonthWeek = { weekNumber: -1, dates: [] };
   for (let i = 0; i <= diff; i++) {
     const date = addDays(monthStart, i);
-    let week = getWeek(date, { locale });
-    if (week === 1 && getMonth(date) === 11) {
-      week = 53;
-    }
-    const weekStr: string = week.toString();
+    // const year = getYear(date, { locale })
```
Comment (gpbl) [inline packages/react-day-picker/src/components/Table/utils/getWeeks.ts:35]:
```suggestion
```

File: packages/react-day-picker/src/components/Table/Table.test.tsx [hunk 6]
```diff
@@ -0,0 +1,43 @@
+import * as React from 'react';
+
+import { enGB, enUS } from 'date-fns/locale';
+import { customRender, PageObjects } from 'test';
+
+import { Table } from './Table';
+
+const po = new PageObjects(new Date());
```
Comment (gpbl) [inline packages/react-day-picker/src/components/Table/Table.test.tsx:8]: We need to freeze date with timekeeper: check out how the test is implemented here: https://github.com/gpbl/react-day-picker/blob/95ec208ccbf97f4c440a196c24c809219d5d7bd3/packages/react-day-picker/src/components/Caption/Caption.test.tsx#L17-L18

File: packages/react-day-picker/src/components/Table/Table.test.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+describe('It should render weeks at the boundaries of years correctly', () => {
```
Comment (gpbl) [inline packages/react-day-picker/src/components/Table/Table.test.tsx:10]: I'm not sure this is the test where we should test this behavior. I would add a test to `getWeeks` instead.

File: packages/react-day-picker/src/components/Table/Table.test.tsx (hunk 2 above)
Comment (gpbl) [reply packages/react-day-picker/src/components/Table/Table.test.tsx:41]: @tjfred35 are you sure from the screen above the week numbers are correct? For December 2021 , my calendar app shows 48–1 for (week starting from Monday) or 47–52 (week starting from Sunday). But never the first week is 49. We should double check this...

File: packages/react-day-picker/src/components/Table/utils/getWeeks.ts [hunk 8]
```diff
@@ -14,51 +13,44 @@ import { getOutsideEndDays } from './getOutsideEndDays';
...
+): MonthWeek[] {
   const monthStart = startOfMonth(month);
   const monthEnd = endOfMonth(month);
 
   const diff = differenceInCalendarDays(monthEnd, monthStart);
 
-  const weeks: MonthWeeks = {};
-  let lastWeekStr = '';
+  const weeks: MonthWeek[] = [];
+  let lastWeek: MonthWeek = { weekNumber: -1, dates: [] };
```
Comment (gpbl) [inline packages/react-day-picker/src/components/Table/utils/getWeeks.ts:32]: 🤔 this shouldn't be:

```suggestion
  let lastWeek: MonthWeek = { weekNumber: 0, dates: [] };
```

**Review by gpbl — COMMENTED** [review]
Hey thanks for your contribution! Week numbers are a beast. They depends from the first day of the week or the locale.

- are the week number still correct? My calendar app says other numbers but it maybe because locale / first day of week
- align the test of `Table.tsx` to our pattern: for example, [Caption.test.tsx](https://github.com/gpbl/react-day-picker/blob/95ec208ccbf97f4c440a196c24c809219d5d7bd3/packages/react-day-picker/src/components/Caption/Caption.test.tsx)
  - in particular make sure to freeze the time
- i think if we change `getWeeks` we must add a test for that file instead of testing table.

File: packages/react-day-picker/src/components/Table/utils/getWeeks.ts (hunk 3 above)
Comment (gpbl) [reply packages/react-day-picker/src/components/Table/utils/getWeeks.ts:18]: I think it is a pretty nice solution. I wasn't sure at first about the `Object`, which i think I choose for performance reasons.

File: packages/react-day-picker/src/components/Table/Table.test.tsx (hunk 2 above)
Comment (tjfred35) [reply packages/react-day-picker/src/components/Table/Table.test.tsx:41]: Week numbers are wild, I guess! On the published beta in codesandbox I can get the following

**locale={enUS}**
Screen Shot 2021-11-04 at 8 34 40 AM

**locale={enGB}**
Screen Shot 2021-11-04 at 8 37 53 AM

https://codesandbox.io/s/daypicker-basics-customization-weeknumber-example-forked-k9o5n

What locale / configuration do you use to get 47-52?

File: packages/react-day-picker/src/components/Table/Table.test.tsx (hunk 6 above)
Comment (tjfred35) [reply packages/react-day-picker/src/components/Table/Table.test.tsx]: Good to know. I ended up moving the tests to getWeeks.test.ts where they got much simpler -- no react necessary!

File: packages/react-day-picker/src/components/Table/Table.test.tsx (hunk 7 above)
Comment (tjfred35) [reply packages/react-day-picker/src/components/Table/Table.test.tsx]: Good idea. That cleaned things up considerably

File: packages/react-day-picker/src/components/Table/utils/getWeeks.ts (hunk 8 above)
Comment (tjfred35) [reply packages/react-day-picker/src/components/Table/utils/getWeeks.ts]: I wanted this initial state to be null, but then I didn't want to deal with null later -- this initial state is never added to the list, and so I wanted to be good and sure the week number was invalid. I changed it to zero, based on your suggestion, but I have assumed zero is never valid.

**Review by tjfred35 — COMMENTED** [review]
Thanks for the review @gpbl. I think this makes things much cleaner, at least as far as the tests are concerned.

File: packages/react-day-picker/src/components/Table/utils/getWeeks.test.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
+import { enGB, enUS } from 'date-fns/locale';
+import { getWeeks } from './getWeeks';
+
```
Comment (gpbl) [inline packages/react-day-picker/src/components/Table/utils/getWeeks.test.ts:3]: We need to "freeze the time" here as well (like in `Caption` with the tk package).

File: packages/react-day-picker/src/components/Table/utils/getWeeks.test.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+describe('It should render weeks at the boundaries of years correctly', () => {
+  test('If the first days of the month start at week 52 from the previous year, week 52 should be rendered on the 1st row', () => {
```
Comment (gpbl) [inline packages/react-day-picker/src/components/Table/utils/getWeeks.test.ts:5]: We should update the tests to follow this format:

```ts
describe('when the first days of the month start at week 52 from the previous year', () => {
  /* some setup here */
  test('the 1st row should render 52', () => {
    expect(); // etc
```

**Review by gpbl — CHANGES_REQUESTED** [review]
LGTM - could we improve the test as suggested? Eg the timekeeper. 

Thanks!

File: packages/react-day-picker/src/components/Table/utils/getWeeks.test.ts (hunk 10 above)
Comment (tjfred35) [reply packages/react-day-picker/src/components/Table/utils/getWeeks.test.ts:5]: I'll try to have just one expect per test

File: packages/react-day-picker/src/components/Table/utils/getWeeks.test.ts (hunk 9 above)
Comment (tjfred35) [reply packages/react-day-picker/src/components/Table/utils/getWeeks.test.ts:3]: I think we may only need to freeze time in the case that we are testing code that uses `new Date()`.

File: packages/react-day-picker/src/components/Table/Table.test.tsx (hunk 2 above)
Comment (gpbl) [reply packages/react-day-picker/src/components/Table/Table.test.tsx]: > What locale / configuration do you use to get 47-52?

Interesting. The "Region" in my Mac OS is set to Spain. Indeed the week number changes. If there are issues we will find them out but it looks like it is expected :)

File: packages/react-day-picker/src/components/Table/utils/getWeeks.ts (hunk 8 above)
Comment (gpbl) [reply packages/react-day-picker/src/components/Table/utils/getWeeks.ts]: Yeah maybe a reduce there would have worked better.

File: packages/react-day-picker/src/components/Table/utils/getWeeks.test.ts (hunk 9 above)
Comment (gpbl) [reply packages/react-day-picker/src/components/Table/utils/getWeeks.test.ts:3]: Would effects happen even if the `new Date()` is used in the test? Anyway thanks for the heads up!

File: packages/react-day-picker/src/components/Table/utils/getWeeks.test.ts (hunk 10 above)
Comment (gpbl) [reply packages/react-day-picker/src/components/Table/utils/getWeeks.test.ts]: Yeah! We should improve the test to be more descriptive, we can do that during some clean-up…

**Review by gpbl — APPROVED** [review]
